octoprint_authldap/__init__.py

```diff
@@ -0,0 +1,271 @@
+
+	def changeUserPassword(self, username, password):
+		# Changing password of LDAP users is not allowed
+		if self.findLDAPUser(username) is None:
+			return self._localUserManager.changeUserPassword(username, password)
+		else:
+			self._logger.error("User is not allowed to change the Password.")
+
+	def getLDAPClient(self):
+		self._logger.debug("Creating LDAP Client")
+		ldap_server=settings().get(["plugins", "authldap", "ldap_uri"])
+		self._logger.debug("LDAP URL %s" % ldap_server)
+		if not ldap_server:
+			settings().remove(["accessControl", "userManager"])
+			self._logger.debug("UserManager: %s" % settings().get(["accessControl", "userManager"]))
+			raise Exception("LDAP conf error, server is missing")
+
+		connection = ldap.initialize(ldap_server)
+		connection.set_option(ldap.OPT_REFERRALS,0)
+		self._logger.debug("LDAP initialized")
+
+		if settings().get(["plugins", "authldap", "ldap_method"]) == 'TLS':
+			self._logger.debug("LDAP is using TLS, setting ldap options...")
+			ldap_verifypeer = settings().get(
+				["plugins", "authldap", "ldap_tls_reqcert"])
+			verifypeer = ldap.OPT_X_TLS_HARD
+			if ldap_verifypeer == 'NEVER':
+				verifypeer = ldap.OPT_X_TLS_NEVER
+			elif ldap_verifypeer == 'ALLOW':
+				verifypeer = ldap.OPT_X_TLS_ALLOW
+			elif ldap_verifypeer == 'TRY':
+				verifypeer = ldap.OPT_X_TLS_TRY
+			elif ldap_verifypeer == 'DEMAND':
+				verifypeer = ldap.OPT_X_TLS_DEMAND
+			# elif ldap_verifypeer == 'HARD':
+			#   verifypeer = ldap.OPT_X_TLS_HARD
+			connection.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, verifypeer)
+			try:
+				connection.start_tls_s()
+				self._logger.error("TLS connection established.")
+			except:
+				self._logger.error("Error initializing tls connection")
+				pass
+		self._logger.error("Finished creating connection")
+		return connection
+
+	def escapeLDAP(self, str):
+		reservedStrings = ['+', '=', '\\', '\r',
+						   '\n', '#', ',', '>', '<', '"', ';']
+		for ch in reservedStrings:
+			if ch in str:
+				str = str.replace(ch, '\\' + ch)
+		return str
+
+	# Softwareupdate hook
+
+	def get_update_information(self):
+		return dict(
+			authldap=dict(
+				displayName="AuthLDAP",
+				displayVersion=self._plugin_version,
+
+				# version check: github repository
+				type="github_release",
+				user="gillg",
+				repo="OctoPrint-LDAP",
+				current=self._plugin_version,
+
+				# update method: pip
+				pip=("https://github.com"
+					 "/gillg/OctoPrint-LDAP/archive/{target_version}.zip")
+			)
+		)
+
+	# UserManager hook
+
+	def ldap_user_factory(components, settings, *args, **kwargs):
+		return LDAPUserManager()
+
+	# SettingsPlugin
+
+	def get_settings_defaults(self):
+		return dict(
+			active=False,
+			ldap_uri=None,
+			ldap_search_base=None,
+			ldap_query=None,
+			ldap_method=None,
+			auto_activate=False,
+			roles=None,
+			ldap_tls_reqcert=None,
+			ldap_bind_user=None,
+			ldap_bind_password = None
+		)
+
+	def on_settings_save(self, data):
```

Comment:
This part only registers/unregisters the UserManager in OctoPrint. A better solution would be to directly write into the accessControl.userManager setting, this is only quick and dirty, but redundant...